src/wp-includes/html-api/class-wp-html-processor.php

```diff
@@ -1165,6 +1254,143 @@ private function step_in_body() {
+			/*
+			 * > A start tag whose tag name is "button"
+			 */
+			case '+BUTTON':
+				if ( $this->state->stack_of_open_elements->has_element_in_scope( 'BUTTON' ) ) {
+					// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
+					$this->generate_implied_end_tags();
+					$this->state->stack_of_open_elements->pop_until( 'BUTTON' );
+				}
+
+				$this->reconstruct_active_formatting_elements();
+				$this->insert_html_element( $this->state->current_token );
+				$this->state->frameset_ok = false;
+
+				return true;
+
+			/*
+			 * > An end tag whose tag name is one of: "address", "article", "aside", "blockquote",
+			 * > "button", "center", "details", "dialog", "dir", "div", "dl", "fieldset",
+			 * > "figcaption", "figure", "footer", "header", "hgroup", "listing", "main",
+			 * > "menu", "nav", "ol", "pre", "search", "section", "summary", "ul"
+			 */
+			case '-ADDRESS':
+			case '-ARTICLE':
+			case '-ASIDE':
+			case '-BLOCKQUOTE':
+			case '-BUTTON':
+			case '-CENTER':
+			case '-DETAILS':
+			case '-DIALOG':
+			case '-DIR':
+			case '-DIV':
+			case '-DL':
+			case '-FIELDSET':
+			case '-FIGCAPTION':
+			case '-FIGURE':
+			case '-FOOTER':
+			case '-HEADER':
+			case '-HGROUP':
+			case '-LISTING':
+			case '-MAIN':
+			case '-MENU':
+			case '-NAV':
+			case '-OL':
+			case '-PRE':
+			case '-SEARCH':
+			case '-SECTION':
+			case '-SUMMARY':
+			case '-UL':
+				if ( ! $this->state->stack_of_open_elements->has_element_in_scope( $token_name ) ) {
+					// @todo Report parse error.
+					// Ignore the token.
+					return $this->step();
+				}
+
+				$this->generate_implied_end_tags();
+				if ( ! $this->state->stack_of_open_elements->current_node_is( $token_name ) ) {
+					// @todo Record parse error: this error doesn't impact parsing.
+				}
+				$this->state->stack_of_open_elements->pop_until( $token_name );
+				return true;
+
+			/*
+			 * > An end tag whose tag name is "form"
+			 */
+			case '-FORM':
+				if ( ! $this->state->stack_of_open_elements->contains( 'TEMPLATE' ) ) {
+					$node                      = $this->state->form_element;
+					$this->state->form_element = null;
+
+					/*
+					 * > If node is null or if the stack of open elements does not have node
+					 * > in scope, then this is a parse error; return and ignore the token.
+					 *
+					 * @todo It's necessary to check if the form token itself is in scope, not
+					 *       simply whether any FORM is in scope.
+					 */
+					if (
+						null === $node ||
+						! $this->state->stack_of_open_elements->has_element_in_scope( 'FORM' )
+					) {
+						// Parse error: ignore the token.
+						return $this->step();
+					}
+
+					$this->generate_implied_end_tags();
+					if ( $node !== $this->state->stack_of_open_elements->current_node() ) {
+						// @todo Indicate a parse error once it's possible. This error does not impact the logic here.
+					}
+
+					$this->state->stack_of_open_elements->remove_node( $node );
```

Comment:
> The </form> close tag should be completely ignored

Why do you say this?

> I don't think we should ever have cases where we remove a node from the middle of the stack like we do here.

This happens after generating the implied end tags, thought that doesn't impact the `SPAN`.

From how I read the spec, it very much _should_ remove the node mid-stack _when we have a parse error that there are unclosed elements_.

This happens during the _adoption agency algorithm_ as well.

Reply:
<img width="593" alt="Screenshot 2024-07-05 at 10 06 39 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/79a7537a-fe7b-4489-b0d6-12ee967aae86">

I'm not sure why we get `after` twice, but it's not confusing I think because of the statefulness of this algorithm. the text nodes are still children of the `SPAN`, but by the time they get added, the `FORM` is no longer open, so it doesn't show up in the breadcrumbs.

we'd have to remember the stack of open elements of the parent when adding the child 🤔

Reply:
I've added an abort condition in [917061b](https://github.com/WordPress/wordpress-develop/pull/6972/commits/917061b9d6369a5711eb16ac2243536bcd112ce0)

the new traversal is this

<img width="607" alt="Screenshot 2024-07-05 at 10 23 00 AM" src="https://github.com/WordPress/wordpress-develop/assets/5431237/fc71df00-dfc0-4e30-b769-6ec94dbaeb78">

we can deal with this particular error condition later

Reply:
You're right, the form close tag is not ignored, but it does take the form off the stack of open elements _out of order_. It's no longer open, but it's children remain open.

This is easier to see in an example you shared elsewhere:

```html
<form>
  <div>In the form</form> still in the form
  </div>
Not in the form
```

HTML API doesn't represent this well now because we treat the stack of open elements as if it were the path up to the document root when constructing breadcrumbs. Often, that works, but not always.

---

I've explored adding Node classes a few times. There are a few situations that might help, and this is one of them. If we worked with Nodes that kept a reference to their parent node, we could construct breadcrumbs from that list. Then when nodes come off the stack of open elements and have no children referencing them, they can be garbage collected.

Reply:
> I've added an abort condition in [917061b](https://github.com/WordPress/wordpress-develop/pull/6972/commits/917061b9d6369a5711eb16ac2243536bcd112ce0)

That's a good approach for now 👍

Reply:
I'm interested in the forgetful nodes idea. Something good to explore down the line.